File: parallelism.cpp
#include <iostream>
#include <Windows.h>
#include <thread>
#include <functional>
#include <mutex>

using namespace std;


typedef function<void()> Callback;
void doit(Callback* cb) {
	/* check cpu parallelism
	while (true){}
	//*/

	Sleep(1000);
	cout << "並行" << endl;
	(*cb)();
}


int main()
{	
	auto threads = 5;
	int count = 0;
	auto mtx = new mutex();

	auto listener = new Callback([&]() {
		mtx->lock();

		++count;
		Sleep(1000);
		cout << count << endl;

		mtx->unlock();
	});
	
	for (auto i = 0; i < threads;++i) {
		auto t = new thread(doit, listener);
	}

	while (threads != count) Sleep(4000);
}

File: parallelism.go
package main

import (
	"fmt"
	"sync"
	"time"
)

func main() {
	threads :=5
	//var wg *sync.WaitGroup=&sync.WaitGroup{}
	wg :=&sync.WaitGroup{}
	wg.Add(threads)

	lobj :=&sync.Mutex{}
	count:=0
	listener:=func() {
		lobj.Lock()

		count++
		time.Sleep(1000*time.Millisecond)
		fmt.Println(count)

		lobj.Unlock()
	}

	for i:=0;i< threads;i++ {
		go doit(wg, listener)
	}

	wg.Wait()
}

func doit(wg *sync.WaitGroup, cb func())  {
	defer func() {
		//fmt.Println("-1")
		wg.Done()
	}()

	/* check cpu parallelism
	for true {}
	// */

	time.Sleep(1000*time.Millisecond)
	fmt.Println("並行")
	cb()
}